src/components/Router/HistoryRouter.tsx

```diff
@@ -1,173 +1,223 @@
-import type { ExtendedStepTypes, FlowVariants } from '~types/commons'
+import type { ErrorCallback, ParsedError } from '~types/api'
+import type { FlowVariants, FormattedError } from '~types/commons'
 import type { CaptureKeys } from '~types/redux'
 import type {
-  ComponentStep,
   ChangeFlowProp,
+  ComponentStep,
+  HistoryLocationState,
   HistoryRouterProps,
 } from '~types/routers'
 import type { SdkResponse } from '~types/sdk'
-import type { DocumentTypes } from '~types/steps'
+import type { DocumentTypes, StepConfig } from '~types/steps'
+import { useCallback, useEffect, useState } from 'preact/hooks'
+import { useHistory } from './useHistory'
 
-type FormattedError = {
-  type: 'expired_token' | 'exception'
-  message: string
-}
+type HistoryRouterState = {
+  initialStep: number
+} & HistoryLocationState
+
+const formattedError = ({ response, status }: ParsedError): FormattedError => {
+  const errorResponse = response.error || response || {}
 
-type HistoryLocationState = {
-  step: number
-  flow: FlowVariants
+  const isExpiredTokenError =
+    status === 401 && errorResponse.type === 'expired_token'
+  const type = isExpiredTokenError ? 'expired_token' : 'exception'
+  // `/validate_document` returns a string only. Example: "Token has expired."
+  // Ticket in backlog to update all APIs to use signature similar to main Onfido API
+  const message = errorResponse.message || response.message || 'Unknown error'
+  return { type, message }
 }
 
-type State = {
-  initialStep: number
-} & HistoryLocationState
+export const HistoryRouter = (props: HistoryRouterProps) => {
+  const {
+    currentStepType,
+    actions,
+    isNavigationDisabled,
+    onFlowChange,
+    captures,
+    options,
+    documentType,
+    step: defaultStep,
+    stepIndexType,
+    deviceHasCameraSupport,
+    options: { mobileFlow, useMemoryHistory },
+    useSteps,
+  } = props
+
+  const getComponentsList = (
+    steps: StepConfig[],
+    flow?: FlowVariants
+  ): ComponentStep[] => {
+    if (!steps) {
+      throw new Error('steps not provided')
+    }
 
-const findFirstIndex = (
-  componentsList: ComponentStep[],
-  clientStepIndex: number
-) => componentsList.findIndex(({ stepIndex }) => stepIndex === clientStepIndex)
+    return buildComponentsList({
+      flow: flow || state.flow,
+      documentType,
+      steps,
+      mobileFlow,
+      deviceHasCameraSupport,
+    })
+  }
 
-export default class HistoryRouter extends Component<
-  HistoryRouterProps,
-  State
-> {
-  private history:
-    | MemoryHistory<HistoryLocationState>
-    | History<HistoryLocationState>
-  private unlisten: () => void
+  const { history, push } = useHistory(({ state: historyState }) => {
+    if (!historyState) {
+      return
+    }
 
-  constructor(props: HistoryRouterProps) {
-    super(props)
+    const { step } = getComponentsList(steps)[historyState.step]
 
-    const componentsList = this.getComponentsList('captureSteps', this.props)
+    if (step.skip) {
```

Comment:
Please enlighten me about the `.skip` property.

Reply:
We need to hide User Consent step when user gave its consent. So I added a generic way to skip a Step from routing and history.